pkg/controller/smart-clone-controller.go

```diff
@@ -0,0 +1,387 @@
+				return true
+			}
+			if snapshotToDelete != nil {
+				klog.V(3).Infof("ProcessNextItem snapshotName: %s", snapshotName)
+				err = c.csiClientSet.SnapshotV1alpha1().VolumeSnapshots(ns).Delete(snapshotName, &metav1.DeleteOptions{})
+				if err != nil {
+					klog.Errorf("error deleting snapshot for smart-clone %q: %v", key, err)
+					return true
+				}
+				klog.V(3).Infof("Snapshot deleted: %s", snapshotName)
+
+				dataVolume, err := c.dataVolumeLister.DataVolumes(snapshot.Namespace).Get(snapshot.Name)
+				if err != nil {
+					return true
+				}
+
+				// Update DV phase and emit PVC in progress event
+				c.updateSmartCloneStatusPhase(cdiv1.Succeeded, dataVolume, pvc)
+			}
+		}
+	} else if snapshot != nil {
+		err := c.syncSnapshot(key.(string))
+		if err != nil {
+			klog.Errorf("error processing snapshot %q: %v", key, err)
+			return true
+		}
+	}
+	return c.forgetKey(key, fmt.Sprintf("ProcessNextSnapshotItem: processing pvc/snapshot %q completed", key))
+}
+
+func (c *SmartCloneController) syncSnapshot(key string) error {
+	snapshot, exists, err := c.snapshotFromKey(key)
+	if err != nil {
+		return err
+	} else if !exists {
+		c.snapshotExpectations.DeleteExpectations(key)
+	}
+
+	if snapshot == nil {
+		return nil
+	}
+	_, ok := snapshot.Annotations[AnnSmartCloneRequest]
+	if !ok {
+		//ignoring snapshot, not created by DataVolume Controller
+		return nil
+	}
+
+	snapshotReadyToUse := snapshot.Status.ReadyToUse
+	klog.V(3).Infof("Snapshot \"%s/%s\" - ReadyToUse: %t", snapshot.Namespace, snapshot.Name, snapshotReadyToUse)
+	if !snapshotReadyToUse {
+		//set Datavolume phase to Scheduled
+		return nil
+	}
+
+	return c.processNextSnapshotItem(snapshot)
+}
+
+// Create the cloning source and target pods based the pvc. The pvc is checked (again) to ensure that we are not already
+// processing this pvc, which would result in multiple pods for the same pvc.
+func (c *SmartCloneController) processNextSnapshotItem(snapshot *csisnapshotv1.VolumeSnapshot) error {
+	snapshotKey, err := cache.MetaNamespaceKeyFunc(snapshot)
+	if err != nil {
+		return err
+	}
+
+	// expectations prevent us from creating multiple pods. An expectation forces
+	// us to observe a pod's creation in the cache.
+	needsSync := c.snapshotExpectations.SatisfiedExpectations(snapshotKey)
+
+	if needsSync {
+		err := c.initializeExpectations(snapshotKey)
+		if err != nil {
+			return err
+		}
+	}
+
+	dataVolume, err := c.dataVolumeLister.DataVolumes(snapshot.Namespace).Get(snapshot.Name)
+	if err != nil {
+		return err
+	}
+
+	// Update DV phase and emit PVC in progress event
+	c.updateSmartCloneStatusPhase(SmartClonePVCInProgress, dataVolume, nil)
+
+	newPvc := newPvcFromSnapshot(snapshot, dataVolume)
+	if newPvc == nil {
+		klog.Errorf("error creating new pvc from snapshot object")
+		return nil
+	}
+
+	c.snapshotExpectations.ExpectCreations(snapshotKey, 1)
+	_, err = c.clientset.CoreV1().PersistentVolumeClaims(snapshot.Namespace).Create(newPvc)
+	if err != nil {
```

Comment:
We appear to be assuming that the PVC definitely does not exist.  But I think that is not guaranteed to be the case.  I think it is possible for our code to try to create twice (pvc exists but is not bound).  Also, what if the pvc already happens to exist by some other means ?  Do we delete/recreate?  Or put the datavolume in a failed state?  Or just keep retrying until it does not exist (current situation).

Reply:
I don't know exactly when the snapshot status transitions to/from `ReadyToUse` but it appears that `syncSnapShot` may called when the pvc exists but is not yet bound.  Is the snapshot status guaranteed to not be `ReadyToUse` at that point?